source/common/ext_authz/ext_authz_impl.cc

```diff
@@ -0,0 +1,241 @@
+      }
+    } else {
+      principal = ssl->uriSanLocalCertificate();
+
+      if (principal.empty()) {
+        principal = ssl->subjectLocalCertificate();
+      }
+    }
+  }
+  peer->set_principal(principal);
+
+  if (!service.empty()) {
+    peer->set_service(service);
+  }
+
+  return peer;
+}
+
+std::unique_ptr<AttributeContext_Peer>
+ExtAuthzCheckRequestGenerator::getConnectionPeer(const Network::Connection& connection,
+                                                 const std::string& service, const bool local) {
+  return getConnectionPeer(&connection, service, local);
+}
+
+const std::string ExtAuthzCheckRequestGenerator::getProtocolStr(const Envoy::Http::Protocol& p) {
+
+  switch (p) {
+  case Envoy::Http::Protocol::Http10:
+    return std::string("Http1.0");
+  case Envoy::Http::Protocol::Http11:
+    return std::string("Http1.1");
+  case Envoy::Http::Protocol::Http2:
+    return std::string("Http2");
+  default:
+    break;
+  }
+  return std::string("unknown");
+}
+
+Envoy::Http::HeaderMap::Iterate
+ExtAuthzCheckRequestGenerator::fillHttpHeaders(const Envoy::Http::HeaderEntry& e, void* ctx) {
+  Envoy::Protobuf::Map<::std::string, ::std::string>* mhdrs =
+      static_cast<Envoy::Protobuf::Map<::std::string, ::std::string>*>(ctx);
+  (*mhdrs)[std::string(e.key().c_str(), e.key().size())] =
+      std::string(e.value().c_str(), e.value().size());
+  return Envoy::Http::HeaderMap::Iterate::Continue;
+}
+
+std::string ExtAuthzCheckRequestGenerator::getHeaderStr(const Envoy::Http::HeaderEntry* entry) {
+  if (entry) {
+    return std::string(entry->value().c_str(), entry->value().size());
+  }
+  return "";
+}
+
+std::unique_ptr<AttributeContext_Request> ExtAuthzCheckRequestGenerator::getHttpRequest(
+    const Envoy::Http::StreamDecoderFilterCallbacks* callbacks,
+    const Envoy::Http::HeaderMap& headers) {
+
+  auto httpreq = new AttributeContext_HttpRequest();
+
+  // Set id
+  // The streamId is not qualified as a const. Although it is as it does not modify the object.
+  Envoy::Http::StreamDecoderFilterCallbacks* sdfc =
+      const_cast<Envoy::Http::StreamDecoderFilterCallbacks*>(callbacks);
+  httpreq->set_id(std::to_string(sdfc->streamId()));
+
+  // Set method
+  httpreq->set_method(getHeaderStr(headers.Method()));
+  // Set path
+  httpreq->set_path(getHeaderStr(headers.Path()));
+  // Set host
+  httpreq->set_host(getHeaderStr(headers.Host()));
+  // Set scheme
+  httpreq->set_scheme(getHeaderStr(headers.Scheme()));
+
+  // Set size
+  // need to convert to google buffer 64t;
+  httpreq->set_size(sdfc->requestInfo().bytesReceived());
+
+  // Set protocol
+  httpreq->set_protocol(getProtocolStr(sdfc->requestInfo().protocol().value()));
+
+  // Fill in the headers
+  auto mhdrs = httpreq->mutable_headers();
+  headers.iterate(fillHttpHeaders, mhdrs);
+
+  auto req = std::make_unique<AttributeContext_Request>();
+
+  req->set_allocated_http(httpreq);
+
+  return req;
+}
```

Comment:
I think it would be instructive to compare with https://github.com/envoyproxy/envoy/blob/master/source/common/access_log/grpc_access_log_impl.cc#L251. There is a lot of overlap between the access log proto data and construction and the auth request.

As a non-actionable reflection, it would have been great if we could have converged the common objects for both access logs and auth, but I don't think this is going to happen (@mattklein123).

Reply:
@htuch This really depends on how big the scale we are targeting. If we target millions of services, billions of requests per second, we can surely refactor the code, but it will take quite a bit work.

The general approach would be something like this:
- Reach an agreement on the standard Envoy attributes, and define them as a proto like AttributeContext.
- Envoy natively generate the standard attributes.
- Design the standard logs, metrics, events, traces , and other telemetry schemas.
- Generate standard telemetry data based on the standard attributes.
- Optional. Generate legacy telemetry data, such as access logs, from the standard attributes.

New users of Envoy won't need the legacy telemetry data, so it would just be disabled by a flag. We can delete the legacy code after a year, or just keep them on the side.

This refactoring would only be useful if we can agree on a rich set of standard telemetry data that appeal to wide range of users, so a customer can correlate telemetry data across services written by different vendors.

Reply:
@htuch Ref to the first paragraph, are you suggesting that I write this in a way that it's done in https://github.com/envoyproxy/envoy/blob/master/source/common/access_log/grpc_access_log_impl.cc#L251.
I am happy to do that but just want to confirm.

Separately, thanks for all the other review comments. I am working on making those changes.

Reply:
@wora agreed, I'm just waxing philosophical. 

@saumoh just suggesting you give that code a pass and see if there's anything in there that your code would benefit from following.